src/main/scala/edu/cmu/cs/obsidian/typecheck/Checker.scala

```diff
@@ -1921,7 +1919,7 @@ class Checker(unmodifiedTable: SymbolTable, verbose: Boolean = false) {
             checkStatementSequence(constr, initContext, constr.body)
 
         val expectedThisType =
-            OwnedRef(table, NoPathType(simpleOf(table, table.name, constr.ensuresState)))
+            OwnedRef(table, NoPathType(simpleOf(table, table.name, constr.endsInState)))
```

Comment:
This will take the least upper bound of the two types, right? We want a more specific union type instead.